### PR TITLE
feat(lsp): Split out a "get qf items" func from set_loclist

### PR DESCRIPTION
**Pull request by mfussenegger**
Makes it easier to re-use the logic to populate the quickfix list
instead of the location list.

### CODE REVIEW

File: runtime/lua/vim/lsp/diagnostic.lua
Comment (mfussenegger) [inline runtime/lua/vim/lsp/diagnostic.lua:1211]: I'm not sure about the name. Happy to change it to something else if there's a good suggestion.

Comment (clason) [reply runtime/lua/vim/lsp/diagnostic.lua:1211]: Maybe something like `locations_to_items` would be more descriptive? And possibly put it in `util.lua`?

As an aside, I see the LSP API as basically having three layers:
1. the server-facing layer, which in naming, data structures and handling should follow LSP terminology as closely as possible (talking about `Locations` and `Tokens` and whatnot) -- these are functions in `lsp.lua` etc.
2. the vim-facing layer, which should "speak vim" -- `clist`, `loclist` etc. -- these are functions in `buf.lua` and `diagnostic.lua`
3. a set of (exposed) utility functions that translate between the two and live in `util.lua`

Comment (mfussenegger) [reply runtime/lua/vim/lsp/diagnostic.lua:1211]: Yes I was considering a `diagnostic_to_items` in `util.lua` first. It's also what I have in my dotfiles https://github.com/mfussenegger/dotfiles/blob/579ab91da5b14ba20766f2acc11d855a56aa0a5f/vim/.config/nvim/lua/me/lsp/diagnostic.lua#L5-L25 it takes the diagnostic grouped by bufnr and a predicate -  you can use it like this:

```
  local items = diagnostics_to_items(
    vim.lsp.diagnostic.get_all(),
    function(d) return d.severity == vim.lsp.protocol.DiagnosticSeverity.Error end
  )
```

I'd prefer that variant (and trimming the amount of supported options). The reason I went with this `get_as_qfitems` is mostly because of how `set_loclist` already supports a lot of options.

I'll wait and see what the others have to say.

Comment (mjlbach) [reply runtime/lua/vim/lsp/diagnostic.lua:1211]: How would you feel about merging this as-is. Would `diagnostics_to_items` be an ok name here, and then we could migrate the function to a utility function?

Comment (mfussenegger) [reply runtime/lua/vim/lsp/diagnostic.lua]: Pushed a fixup with the rename. Let me know if it is otherwise good to go - I'll squash before it gets merged.

Comment (mfussenegger) [reply runtime/lua/vim/lsp/diagnostic.lua]: Actually, let me try another variant first.

Comment (mfussenegger) [reply runtime/lua/vim/lsp/diagnostic.lua]: I pushed another fixup - adding a `util.diagnostics_to_items`. It also includes a fix for `set_loclist` (previously if `workspace` was `true`, the `client_id` option was ignored)

Let me know which variant you prefer, the initial variant in the first commit - or the one in the last commit.

File: runtime/lua/vim/lsp/util.lua
Comment (mfussenegger) [inline runtime/lua/vim/lsp/util.lua:1907]: I left out the `character_offset` call because it is wrong. We'd need to convert utf-16 offsets to byte positions if we want to be correct here.

File: runtime/lua/vim/lsp/diagnostic.lua
Comment (mjlbach) [inline runtime/lua/vim/lsp/diagnostic.lua:1187]: Added blankspace?

**Comment by mjlbach** [comment]
Also, realize I didn't answer your question. I think I prefer the latest variant but it's your call.

File: runtime/lua/vim/lsp/diagnostic.lua
Comment (mfussenegger) [reply runtime/lua/vim/lsp/diagnostic.lua]: Removed it

**Comment by mfussenegger** [comment]
> I think I prefer the latest variant but it's your call.

Okay, went with the later version - I also prefer it.

**Review by mjlbach — APPROVED** [review]

**Comment by mjlbach** [comment]
I will merge once tests are green

**Pull request merged**
